### PR TITLE
Add RTL Type To Copter

### PR DESCRIPTION
**Pull request by peterbarker**
Replaces #4723

### CODE REVIEW

File: ArduCopter/Parameters.cpp [hunk 1]
```diff
@@ -918,6 +918,13 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
     // @User: Standard
     AP_GROUPINFO("TUNE_MAX", 32, ParametersG2, tuning_max, 0),
 
+    // @Param: RTL_TYPE
+    // @DisplayName: RTL handling type
+    // @Description: This controls how RTL is handled. A value of zero means a conventional RTL. A value of 1 uses the closest DO_LAND_START marker in the mission waypoints to fly a pre-planned path. If no DO_LAND_START markers a found a conventional RTL is used.
```
Comment (WickedShell) [inline ArduCopter/Parameters.cpp]: Should note that this is using the closest geographic DO_LAND_START to the current vehicle's position not the closest mission planned number.

**Comment by khancyr** [comment]
I am in phone so it is hard to check, but does this work when calling rtl from auto mode? I don't think we are calling mode change

**Comment by peterbarker** [comment]
On Tue, 11 Jun 2019, Pierre Kancir wrote:

> I am in phone so it is hard to check, but does this work when calling rtl
> from auto mode? I don't think we are calling mode change

It substitutes the mode-change-to-RTL - it never actually enters RTL but 
will stay in auto.

That being said, I wasn't setting reason appropriate / skipping the mode 
change if we weren't switching, so I've done that.

**Comment by rmackay9** [comment]
I like the concept of having RTL being to use a predefined mission to get home (although I think really it should trace backwards through the mission) but I'm not too keen on this implementation which leads to the user asking for RTL but actually getting Auto mode.  Worst case it should execute Auto from within RTL mode although then we need to be careful about Auto-within-RTL executing an RTL mission command.

**Comment by rmackay9** [comment]
I'll probably take on restructuring this after we make a bit more progress on the object avoidance stuff.

**Comment by WickedShell** [comment]
@rmackay9 FWIW this is the same behavior plane users get at the moment, and while it seems weird since you are executing commands in auto at this point it is an accurate reflection of the vehicle state (also makes it clear that all your normal auto nudging of stuff like throttle/tuning is active). But just because plane does it one way doesn't mean copter needs to :)

**Comment by Pedals2Paddles** [comment]
- I wonder if this parameter should have three options: Always use normal RTL, Always use do_land_start, Normal RTL except use do_land_start when in Auto Mode only.

- What happens if there is an RTL mission command at the end of the mission (after a do_land_start)?  Would that not put this into an endless loop?

- Is there a sanity check on how far away the a mission can begin?  If an old mission is loaded that is 1000 miles away, and RTL is your RC failsafe, that's going to be a flyaway.  Maybe it should ignore a do_land_start that is more than X distance away from home or from the copter?

- Should this even be part of RTL?  After all, RTL is Return To **L**aunch and this will be landing somewhere, not necessarily at launch..  Maybe RTL should stay RTL, stay simple, and do what it does.  This is basically a shortcut to a point in an auto mission.  That shortcut can be called by a failsafe action an aux switch, a mavlink command, or even be a flight mode that just executes the shortcut into auto mode.

**Comment by rmackay9** [comment]
I agree with @Pedals2Paddles in a couple of ways.  I was thinking a better solution would be:

- add it as a new flight mode perhaps called Return-On-Mission or RTM or something similar
- we should be able to find the path home along the mission path without any special markers in the mission.  We should instead be able to find the closest point to a segment of the mission
- ideally we would handle moving backwards along the mission (i.e. the user has specified the mission in the regular way starting at home)
- if we want to be really advanced we handle taking short-cuts if two segments cross each other although I suspect this would add too much complexity for the benefit it provides

I think the implementation would involve:

- building up a big list of segments by running through the mission from start to finish.
- find the vehicle's closest point on one of the segments
- feed the next segment into the navigation controller until we get home and then trigger the final stages of regular RTL.

**Comment by Pedals2Paddles** [comment]
That would only work if the mission begins with a takeoff command, and the copter actually begins the mission from that takeoff.  If you start the mission with a regular airborne waypoint, that won't get you home. It will get to some random location in the sky instead.

I actually like the do_land_start and onward method.  It allows you to pick a return location other than home if you want, and it allows you to be very specific how you want the copter to approach that landing.  Even if your mission begins with a takeoff command at home, flying the reverse course is not necessarily an ideal approach to landing. In fact I would say it is highly unlikely to be an ideal approach to landing.

Using the do_land_start method as it does now ensures that if this executes, it is taking a course that was specifically intended to be an approach to landing course.  If the mission has no do_land_start, then there is no assurance that the mission has a course intended for an approach to landing.

**Pull request closed without merging**